core/src/game_state.rs

```diff
@@ -1060,8 +1086,40 @@ impl PlayPhase {
             }
         }
 
+        let mut smaller_landlord_team = false;
```

Comment:
I think you actually want

```
let smaller_landlord_team = match self.game_mode {
  GameMode::FindingFriends { num_friends, ref friends } if num_friends > friends.flat_map(|f| f.player_id).collect<HashSet<PlayerID>>().len() => true,
  _ => false,
};
```

In order for this to be correct -- `num_friends` and `friends.len()` are the same, and the interesting bit for how many players are _actually_ on the team is how many unique `friend.player_id`

Reply:
not sure this is correct. I did try something similar to this, and had to back out the changes. The game_mode from the state has the actual current number of friends in num_friends and friends holds the current list of friends, they actually match. the game_mode of the propagated has the setting. I checked using the debugger to confirm as well. Please let me know if I missed anything. Will update the UI.

Reply:
https://github.com/rbtying/shengji/blob/9bb25ba240d903901aa2db82e3bb29e8f3133f32/core/src/game_state.rs#L1683

^ This is where `num_friends` is taken from `propagated` to start the game -- if you set "num_friends" too high, it might have broken your experiment? (in that case, we would correctly want to use the value in `self.game_mode.num_friends`

I realized, actually, that it's insufficient to just check the number of unique player IDs < num_friends, but also that the landlord's ID should not be one of those player IDs. So the comparison needs to account for that as well

If you write a test for this case, I can poke at why it is doing something unexpected?

Reply:
Here how I checked. 
1. played with 4 player
2. landlord select a friend
3. during play, landlord picked himself as friend
4. landlord won the game
5. when I used only state.game_mode, bonus level was not added. I used the debugger to check, num_friends is 1, and friends.len is one as well. 
ah, as I am writing this,  you are right num_friends is the configured number of friends (EXCLUDING the landlord), friends.len is the actual friends list, including the landlord. This should simplify the test!

Reply:
num_friends can never be too high, it is capped at players.len()/2

Reply:
players.len()/2 - 1